### PR TITLE
Datatype Date32, support range 1925 to 2283

### PR DESCRIPTION
**Pull request by liuneng1994**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add a new datatype Date32 (store data as Int32), support date range same with Datetime64
support load parquet date32 to ClickHouse Date32
Add new function toDate32 like toDate
Support some function :

- toYear
- toMonth
- toQuarter
- toDayOfMonth
- toDayOfWeek
- toDayOfYear
- toStartOfDay
- toMonday
- toISOWeek
- toISOYear
- toWeek
- toYearWeek
- toStartOfWeek
- toStartOfMonth
- toStartOfQuarter
- toStartOfYear
- toStartOfISOYear
- toRelativeYearNum
- toRelativeQuarterNum
- toRelativeMonthNum
- toRelativeWeekNum
- toRelativeDayNum
- toRelativeHourNum
- toRelativeMinuteNum
- toRelativeSecondNum
- toYYYYMM
- toYYYYMMDD
- toYYYYMMDDhhmmss
- addSeconds
- addMinutes
- addHours
- addDays
- addWeeks
- addMonths
- addQuarters
- addYears
- subtractSeconds
- subtractMinutes
- subtractHours
- subtractDays
- subtractWeeks
- subtractMonths
- subtractQuarters
- subtractYears

Detailed description / Documentation draft:

I will add more docs in the future

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by liuneng1994** [comment]
retest this please

**Comment by liuneng1994** [comment]
@ kssenii I fix some functional tests.There are still two failure cases, but they seem to have nothing to do with my code. And AST fuzzer failed， I didn't change the wrong line of code. I don't know how to fix it

**Comment by kssenii** [comment]
Ok! I'll try to review as soon as I can (most likely tomorrow).

**Comment by liuneng1994** [comment]
> Ok! I'll try to review as soon as I can (most likely tomorrow).

File: src/IO/ReadHelpers.h [hunk 1]
```diff
@@ -632,6 +632,22 @@ inline ReturnType readDateTextImpl(DayNum & date, ReadBuffer & buf)
...
+    static constexpr bool throw_exception = std::is_same_v<ReturnType, void>;
+
+    LocalDate local_date;
+
+    if constexpr (throw_exception)
+        readDateTextImpl<ReturnType>(local_date, buf);
+    else if (!readDateTextImpl<ReturnType>(local_date, buf))
+        return false;
+
+    date = DateLUT::instance().makeDayNum(local_date.year(), local_date.month(), local_date.day(), -DateLUT::instance().getDaynumOffsetEpoch());
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/IO/ReadHelpers.h:647]: When the parameter is out of rule or out of range, Date32 uses 1925-01-01 as the default value (-DateLUT::instance().getDaynumOffsetEpoch(), -16436) and Date uses 1970-01-01.

File: src/AggregateFunctions/AggregateFunctionSequenceMatch.cpp [hunk 2]
```diff
@@ -63,6 +64,8 @@ AggregateFunctionPtr createAggregateFunctionSequenceBase(
         return std::make_shared<AggregateFunction<DataTypeDateTime::FieldType, Data<DataTypeDateTime::FieldType>>>(argument_types, params, pattern);
     else if (which.isDate())
         return std::make_shared<AggregateFunction<DataTypeDate::FieldType, Data<DataTypeDate::FieldType>>>(argument_types, params, pattern);
+//    else if (which.isDate32())
+//        return std::make_shared<AggregateFunction<DataTypeDate32::FieldType, Data<DataTypeDate32::FieldType>>>(argument_types, params, pattern);
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/AggregateFunctions/AggregateFunctionSequenceMatch.cpp:68]: This function is not supported for now, I will remove the redundant comments

[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/AggregateFunctions/AggregateFunctionSequenceMatch.cpp]: done

File: base/common/DateLUTImpl.h [hunk 3]
```diff
@@ -18,6 +18,7 @@
 
 #define DATE_LUT_MAX (0xFFFFFFFFU - 86400)
 #define DATE_LUT_MAX_DAY_NUM 0xFFFF
+#define DATE_LUT_MAX_EXTEND_DAY_NUM (DATE_LUT_SIZE - 16436)
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply base/common/DateLUTImpl.h:22]: done

File: src/Functions/FunctionDateOrDateTimeAddInterval.h [hunk 4]
```diff
@@ -445,8 +487,15 @@ class FunctionDateOrDateTimeAddInterval : public IFunction
         {
             // TODO (vnemkov): what if there is an overload of Transform::execute() that returns DateTime64 from DateTime or Date ?
             // Shall we use the default scale or one from optional argument ?
-            const auto & datetime64_type = assert_cast<const DataTypeDateTime64 &>(*arguments[0].type);
-            return std::make_shared<DataTypeDateTime64>(datetime64_type.getScale(), extractTimeZoneNameFromFunctionArguments(arguments, 2, 0));
+            if (arguments[0].type->getTypeId() == TypeIndex::DateTime64)
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/Functions/FunctionDateOrDateTimeAddInterval.h]: done

File: base/common/DateLUTImpl.h [hunk 5]
```diff
@@ -270,6 +271,8 @@ class DateLUTImpl
     auto getOffsetAtStartOfEpoch() const { return offset_at_start_of_epoch; }
     auto getTimeOffsetAtStartOfLUT() const { return offset_at_start_of_lut; }
 
+    auto getDaynumOffsetEpoch() const { return daynum_offset_epoch; }
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply base/common/DateLUTImpl.h]: done

**Comment by kssenii** [comment]
I think that here https://github.com/ClickHouse/ClickHouse/blob/9c2aad6c1781b7b4554f792c27089f5f2e4ebcce/base/common/DateLUTImpl.h#L1156-L1164
data types need to be adjusted (hint from fuzzer `../base/common/../common/DateLUTImpl.h:1173:35: runtime error`), the max now will be Int32 after adding Date32, there is no need in int64. Lets make it Int32 in methods which then call inside addMonths. Please, check similar places there.

**Comment by liuneng1994** [comment]
> I think that here
> 
> https://github.com/ClickHouse/ClickHouse/blob/9c2aad6c1781b7b4554f792c27089f5f2e4ebcce/base/common/DateLUTImpl.h#L1156-L1164
> 
> 
> data types need to be adjusted (hint from fuzzer `../base/common/../common/DateLUTImpl.h:1173:35: runtime error`), the max now will be Int32 after adding Date32, there is no need in int64. Lets make it Int32 in methods which then call inside addMonths. Please, check similar places there.

I fix this issue, fuzzer test is ok

File: src/Functions/FunctionDateOrDateTimeAddInterval.h [hunk 6]
```diff
@@ -445,8 +487,15 @@ class FunctionDateOrDateTimeAddInterval : public IFunction
         {
             // TODO (vnemkov): what if there is an overload of Transform::execute() that returns DateTime64 from DateTime or Date ?
             // Shall we use the default scale or one from optional argument ?
-            const auto & datetime64_type = assert_cast<const DataTypeDateTime64 &>(*arguments[0].type);
-            return std::make_shared<DataTypeDateTime64>(datetime64_type.getScale(), extractTimeZoneNameFromFunctionArguments(arguments, 2, 0));
+            if (typeid_cast<const DataTypeDateTime64 *>(arguments[0].type.get()))
```
Comment (kssenii) [inline src/Functions/FunctionDateOrDateTimeAddInterval.h:490]: Comment above can be removed now since it is fixed.

File: src/IO/ReadHelpers.h [hunk 7]
```diff
@@ -632,6 +632,22 @@ inline ReturnType readDateTextImpl(DayNum & date, ReadBuffer & buf)
...
+
+    LocalDate local_date;
+
+    if constexpr (throw_exception)
+        readDateTextImpl<ReturnType>(local_date, buf);
+    else if (!readDateTextImpl<ReturnType>(local_date, buf))
+        return false;
+
+    date = DateLUT::instance().makeDayNum(local_date.year(), local_date.month(), local_date.day(), -DateLUT::instance().getDayNumOffsetEpoch());
+    return ReturnType(true);
```
Comment (kssenii) [inline src/IO/ReadHelpers.h:648]: Please, add this comment 
`When the parameter is out of rule or out of range, Date32 uses 1925-01-01 as the default value (-DateLUT::instance().getDaynumOffsetEpoch(), -16436) and Date uses 1970-01-01.`
into code.

**Review by kssenii — APPROVED** [review]

**Comment by kssenii** [comment]
Also please run `ClickHouse/utils/generate-ya-make/generate-ya-make.sh` script to fix `Yandex third-party checks `.

File: base/common/DateLUTImpl.h [hunk 8]
```diff
@@ -926,15 +930,16 @@ class DateLUTImpl
     {
         if (unlikely(year < DATE_LUT_MIN_YEAR || year > DATE_LUT_MAX_YEAR || month < 1 || month > 12 || day_of_month < 1 || day_of_month > 31))
             return LUTIndex(0);
-
-        return LUTIndex{years_months_lut[(year - DATE_LUT_MIN_YEAR) * 12 + month - 1] + day_of_month - 1};
+        auto year_lut_index = (year - DATE_LUT_MIN_YEAR) * 12 + month - 1;
+        UInt32 index = years_months_lut[year_lut_index].toUnderType() + day_of_month - 1;
+        return LUTIndex{index < DATE_LUT_SIZE ? index : DATE_LUT_SIZE - 1};
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply base/common/DateLUTImpl.h:935]: toLUTIndex doesn't support covert uint32 index to LUTIndex. I use date_lut_mask to protect overflow now

File: src/IO/ReadHelpers.h (hunk 7 above)
Comment (liuneng1994) [reply src/IO/ReadHelpers.h:648]: done

**Comment by liuneng1994** [comment]
> Also please run `ClickHouse/utils/generate-ya-make/generate-ya-make.sh` script to fix `Yandex third-party checks `. It will add some changes to `ya.make` files.

I have run this script, but third-party checks still fail

File: src/Core/ya.make [hunk 9]
```diff
@@ -31,6 +31,10 @@ SRCS(
     MySQL/PacketsProtocolText.cpp
     MySQL/PacketsReplication.cpp
     NamesAndTypes.cpp
+    PostgreSQL/Connection.cpp
+    PostgreSQL/PoolWithFailover.cpp
+    PostgreSQL/Utils.cpp
+    PostgreSQL/insertPostgreSQLValue.cpp
```
Comment (kssenii) [inline src/Core/ya.make:37]: Can you please remove these, `yandex check` fails because of including them. I'll fix them separately.

**Comment by liuneng1994** [comment]
@kssenii failed again...

**Comment by kssenii** [comment]
`Yandex third-party checks (only for Yandex employees)` failure not related to changes, its broken in master: ` Code: 46. DB::Exception: Received from localhost:38565. DB::Exception: Unknown function h3ToChildren: While processing h3ToChildren(599405990164561919, 100).`

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11585.

**Comment by alexey-milovidov** [comment]
@kssenii One bug was found: https://github.com/ClickHouse/ClickHouse/issues/26351
I will revert this PR before the bug will be fixed.

**Comment by liuneng1994** [comment]
I found that my github id disappeared from the code base. .

**Comment by alexey-milovidov** [comment]
@liuneng1994 Your commits are recorded:
```
milovidov@milovidov-desktop:~/work/ClickHouse$ git log | grep 'neng.liu'
Author: neng.liu <1398775315@qq.com>
Author: neng.liu <1398775315@qq.com>
Author: neng.liu <1398775315@qq.com>
Author: neng.liu <1398775315@qq.com>
Author: neng.liu <1398775315@qq.com>
Author: neng.liu <1398775315@qq.com>
Author: Neng Liu <neng.liu@kyligence.io>
Author: Neng Liu <neng.liu@kyligence.io>
Author: neng.liu <1398775315@qq.com>
Author: Neng Liu <neng.liu@kyligence.io>
Author: Neng Liu <neng.liu@kyligence.io>
Author: Neng Liu <neng.liu@kyligence.io>
Author: neng.liu <1398775315@qq.com>
Author: Neng Liu <neng.liu@kyligence.io>
Author: neng.liu <1398775315@qq.com>
Author: Neng Liu <neng.liu@kyligence.io>
Author: neng.liu <1398775315@qq.com>
Author: Neng Liu <neng.liu@kyligence.io>
```

And you will appear in the `system.contributors` table and in the authors in changelog.
The full history of development is kept, including your commits, revert and revert of revert.

**Comment by liuneng1994** [comment]
> @liuneng1994 Your commits are recorded:
> 
> ```
> milovidov@milovidov-desktop:~/work/ClickHouse$ git log | grep 'neng.liu'
> Author: neng.liu <1398775315@qq.com>
> Author: neng.liu <1398775315@qq.com>
> Author: neng.liu <1398775315@qq.com>
> Author: neng.liu <1398775315@qq.com>
> Author: neng.liu <1398775315@qq.com>
> Author: neng.liu <1398775315@qq.com>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: neng.liu <1398775315@qq.com>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: neng.liu <1398775315@qq.com>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: neng.liu <1398775315@qq.com>
> Author: Neng Liu <neng.liu@kyligence.io>
> Author: neng.liu <1398775315@qq.com>
> Author: Neng Liu <neng.liu@kyligence.io>
> ```
> 
> And you will appear in the `system.contributors` table and in the authors in changelog.
> The full history of development is kept, including your commits, revert and revert of revert.

OK, thanks very much

**Pull request merged**